### PR TITLE
feat: op registration during calls

### PR DESCRIPTION
**Pull request by afinch7**
prerequisite for #3372

### CODE REVIEW

File: core/isolate.rs [hunk 1]
```diff
@@ -256,13 +256,18 @@ impl Isolate {
...
   where
     F: Fn(&[u8], Option<PinnedBuf>) -> CoreOp + Send + Sync + 'static,
   {
     self.op_registry.register(name, op)
   }
 
+  /// Returns a thread safe instance of the op registry for this isolate.
+  pub fn thread_safe_op_registry<F>(&self) -> Arc<OpRegistry> {
+    self.op_registry.clone()
+  }
```
Comment (ry) [inline core/isolate.rs]: Is this method necessary? I think you could just make `op_registry` public?

File: core/Cargo.toml [hunk 2]
```diff
@@ -15,7 +15,7 @@ path = "lib.rs"
 
 [dependencies]
 downcast-rs = "1.1.1"
-futures = { version = "0.3", features = [ "thread-pool", "compat" ] }
+futures = { version = "0.3", features = [ "thread-pool", "compat", "io-compat" ] }
```
Comment (ry) [inline core/Cargo.toml]: Is this bit intentional? If so where is it used?

Comment (afinch7) [reply core/Cargo.toml]: I was getting a compile error for http_bench.

File: core/ops.rs [hunk 3]
```diff
@@ -25,17 +27,18 @@ pub type CoreError = ();
...
-type OpDispatcher = dyn Fn(&[u8], Option<PinnedBuf>) -> CoreOp;
+type OpDispatcher =
+  dyn Fn(&[u8], Option<PinnedBuf>) -> CoreOp + Send + Sync + 'static;
 
 #[derive(Default)]
 pub struct OpRegistry {
-  dispatchers: Vec<Box<OpDispatcher>>,
-  name_to_id: HashMap<String, OpId>,
+  dispatchers: RwLock<Vec<Arc<Box<OpDispatcher>>>>,
+  name_to_id: RwLock<HashMap<String, OpId>>,
```
Comment (bartlomieju) [inline core/ops.rs]: 😭 this might impact performance, hot path will require to acquire a read lock everytime

Comment (afinch7) [reply core/ops.rs]: It doesn't look like it did in benchmarks.

Comment (bartlomieju) [reply core/ops.rs]: Yes, but they're flaky... Anyway why do you need to put `dispatchers` behind the lock? In native plugin PR you already have `Arc<Mutex<Isolate>>` to you should be able to register new ops quite easily

Comment (afinch7) [reply core/ops.rs]: In most cases during dispatch we already have a lock on isolate somewhere else.

Comment (bartlomieju) [reply core/ops.rs]: I see... I'm curious to see it combined with native plugins PR

**Comment by afinch7** [comment]
It doesn't look like it messed with the benchmarks much at all. I think it would take something pretty drastic on the rust side to really be noticeable beyond max latency.

**Pull request merged**